### PR TITLE
Feature/hide on click outside

### PR DESCRIPTION
**Pull request by joao-alberto**
Ref #160 

| Example |
| ---- |
| click-outside |

I think this is the best solution, but I'm not the master of jest or enzyme and i'm stuck in this test, can someone help?

[The tests](https://github.com/reakit/reakit/compare/master...joao-alberto:feature/hideOnClickOutside?expand=1#diff-8f9f46188c5c63548c9bc07beb7c6832R65)

Error that i'm am stuck
screen shot 2018-07-31 at 14 37 04

### CODE REVIEW

File: src/components/Hidden/Hidden.js [hunk 1]
```diff
@@ -61,6 +68,14 @@ class Component extends React.Component {
     }
   };
 
+  handleClickOutside = e => {
+    const { className, hide, visible } = this.props;
+
+    if (!e.target.closest(`[class^='${className}']`) && visible && hide) {
```
Comment (foxyblocks) [inline src/components/Hidden/Hidden.js]: I think this won't work if the clicked element has the same class as this one. 

Here is an alternative approach that uses `Node.contains` to perform the check which is slightly more bullet proof: https://codesandbox.io/s/94kwvn31oy

It still has the drawback though that it won't work through [portals](https://reactjs.org/docs/portals.html). If a child component of the <Hidden /> component renders something to a portal, (such as a modal or dropdown or something) then clicks inside the Portal tree will not be considered to be "inside" the <hidden /> components DOM hierarchy and thus will be considered "outside" clicks. I have implemented workarounds to this in a different project but I'm not sure how they should be implemented in this library so that they are easy to use and unsurprising.

Comment (joao-alberto) [reply src/components/Hidden/Hidden.js]: > Here is an alternative approach that uses Node.contains to perform the check which is slightly more bullet proof: https://codesandbox.io/s/94kwvn31oy

Interesting, I didn't realize that will not work in this case, your solution probably is the best

> I have implemented workarounds to this in a different project but I'm not sure how they should be implemented in this library so that they are easy to use and unsurprising.

Can you please share this workarounds with us for discussion and tests?

**Comment by wordofchristian** [comment]
Yes. I’m away from my laptop though. For now I think it’s fine that this implementation doesn’t cover the portal edge case

Christian Schlensker

File: src/components/Hidden/Hidden.js (hunk 1 above)
Comment (foxyblocks) [reply src/components/Hidden/Hidden.js]: Here is the implementation of what we do at bugsnag: 

https://gist.github.com/wordofchristian/78348722edd275880dde54c549a0f155

We call it ClickBoundary. Every ClickBoundary uses context to keep track of any other click boundaries that are descendants of it (context will work across portals). The key is that it only works if the content of the portal is also wrapped in a ClickBoundary, if not, it won't get added to the tree of tracked nodes. So it's not entirely bullet proof from a usage standpoint.

File: src/components/Hidden/__tests__/Hidden-test.js [hunk 2]
```diff
@@ -43,6 +44,29 @@ it("calls hide when press Escape", () => {
   expect(props.hide).toHaveBeenCalledTimes(1);
 });
 
+it("calls hide when click outside", () => {
+  const props = {
+    hide: jest.fn(),
+    hideonClickOutside: true
```
Comment (rthomazel) [inline src/components/Hidden/__tests__/Hidden-test.js]: typo -> `hideOnClickOutside`

File: src/components/Hidden/__tests__/Hidden-test.js [hunk 3]
```diff
@@ -43,6 +44,29 @@ it("calls hide when press Escape", () => {
...
+  };
+
+  const wrapper = shallow(
+    <div>
+      <Hidden {...props} />
+      <button className="any-button" type="button" onClick={jest.fn}>
```
Comment (rthomazel) [inline src/components/Hidden/__tests__/Hidden-test.js]: Did you mean `onClick={jest.fn()}`?

**Comment by Thomazella** [comment]
I tried to get that test to pass but with no luck. Are we sure `ShallowWrapper.simulate` will work to test click outside? Doesn't it just look at the event type `click` and call `onClick`? [see "Common Gotchas"](https://github.com/airbnb/enzyme/blob/master/docs/api/ShallowWrapper/simulate.md)

I'll leave this here in case it helps
```
const component = wrapper.find(Hidden).dive();
const button = wrapper.find(".any-button");
const clickEvent = new MouseEvent("click");

// button.simulate("click");
document.body.dispatchEvent(clickEvent);
expect(props.hide).toHaveBeenCalledTimes(0);
component.setProps({ visible: true });
document.body.dispatchEvent(clickEvent);
// button.simulate("click");
expect(props.hide).toHaveBeenCalledTimes(1);
```

**Comment by joao-alberto** [comment]
@Thomazella thanks for the tips, I'll work on this today

File: src/components/Hidden/__tests__/Hidden-test.js (hunk 2 above)
Comment (joao-alberto) [reply src/components/Hidden/__tests__/Hidden-test.js]: Done

File: src/components/Hidden/__tests__/Hidden-test.js (hunk 3 above)
Comment (joao-alberto) [reply src/components/Hidden/__tests__/Hidden-test.js]: Done

File: src/components/Hidden/__tests__/Hidden-test.js [hunk 4]
```diff
@@ -47,23 +48,16 @@ it("calls hide when press Escape", () => {
...
   const props = {
     hide: jest.fn(),
-    hideonClickOutside: true
+    hideOnClickOutside: true,
+    visible: false
   };
+  const wrapper = mount(<Hidden {...props} />);
+  const clickEvent = new MouseEvent("click");
 
-  const wrapper = shallow(
```
Comment (rthomazel) [inline src/components/Hidden/__tests__/Hidden-test.js]: What is the (practical) difference between `shallow` and `mount` ? I'm curious about it.

Comment (joao-alberto) [reply src/components/Hidden/__tests__/Hidden-test.js]: Shallow is real unit test (isolation, no children render) for this reason the tests were not going
Mount provides a full dom rendering, ideal for use cases where you have components that may interact with DOM APIs

Comment (rthomazel) [reply src/components/Hidden/__tests__/Hidden-test.js]: Thanks for explaining!

**Comment by diegohaz** [comment]
The click outside was conflicting with toggle button. It was hiding the Hidden element (because we're clicking outside it) and quickly showing it again (because we're clicking on the toggle button).

I had to add a `setTimeout` in order to fix it. Now it seems to work well.

Also fixed some typos in the docs and added the prop to Popover example.

**Comment by joao-alberto** [comment]
Thank you guys for the help @wordofchristian @Thomazella @diegohaz :metal:
